File: Java/1382_Balance_a_Binary_Search_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> list = new ArrayList<>();
    public TreeNode balanceBST(TreeNode root) {
        
        addToList(root);
        //System.out.println(list);
        
        TreeNode result = build(list.stream().mapToInt(i->i).toArray());
        
        return result;
    }
    public TreeNode build(int[] arr){
        int n = arr.length;
        if(n == 0){
            return null;
        }
        
        int mid = n/2;
        TreeNode node = new TreeNode(arr[mid]);
        
        int[] left = new int[mid];
        int[] right = new int[n-mid-1];
        
        System.arraycopy(arr, 0, left, 0, mid);
        System.arraycopy(arr, mid+1, right, 0, n-mid-1);
        
        node.left = build(left);
        node.right = build(right);
        
        return node;
    }
    
    
    public void addToList(TreeNode root){
        if(root.left != null)  
            addToList(root.left);
        
        list.add(root.val);
        
        if(root.right != null)
            addToList(root.right);
    }
}

File: CS/1382_Balance_a_Binary_Search_Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    List<int> list = new List<int>();
    TreeNode result = new TreeNode();
    public TreeNode BalanceBST(TreeNode root) {
        treeToList(root);
        
        result = build(0, list.Count-1);
        
        return result;
    }
    private TreeNode build(int l, int r){
        
        if(r-l+1 == 0){
            return null;
        }
        
        TreeNode root = new TreeNode();
        int n = l + (r-l+1)/2;
        root.val = list[n];
        
        root.left = build(l, n-1);
        root.right = build(n+1, r);
        
        return root;
    }
    private void treeToList(TreeNode root){
        if(root == null){
            return;
        }
        treeToList(root.left);
        list.Add(root.val);
        treeToList(root.right);
    }
}
